### PR TITLE
[Sparse] Add Two Sparse Matrices with Different Sparsity, DiagMatrix.dense, Add SparseMatrix and DiagMatrix

### PR DESCRIPTION
**Pull request by mufeili**
### Major
- `A1 + A2`: Add two sparse matrices with different sparsity using PyTorch's implementation for COO matrices. The CSR support of PyTorch is still in beta phase, so I think it's better to wait for a few more versions.
- `D + A` and `A + D` using `A1 + A2`
### Minor
- `DiagMatrix.dense`
- Fix of `SparseMatrix.dense` to handle dtypes other than float

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [ ] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

### CODE REVIEW

File: dgl_sparse/src/elemenwise_op.cc [hunk 1]
```diff
@@ -21,16 +21,14 @@ namespace sparse {
...
-  if (fmt == SparseFormat::kCOO) {
-    return SparseMatrix::FromCOO(A->COOPtr(), value, A->shape());
-  } else if (fmt == SparseFormat::kCSR) {
-    return SparseMatrix::FromCSR(A->CSRPtr(), value, A->shape());
-  } else {
-    return SparseMatrix::FromCSC(A->CSCPtr(), value, A->shape());
-  }
+  auto torch_A = COOToTorchCOO(A->COOPtr(), A->value());
+  auto torch_B = COOToTorchCOO(B->COOPtr(), B->value());
+  auto sum = (torch_A + torch_B).coalesce();
```
Comment (mufeili) [inline dgl_sparse/src/elemenwise_op.cc:27]: `coalesce` is required in order to invoke `.indices()`.

File: tests/pytorch/mock_sparse2/test_elementwise_op.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+from dgl.mock_sparse2 import (add, create_from_coo, create_from_csc,
+                              create_from_csr, diag)
+
+# TODO(#4818): Skipping tests on win.
+if not sys.platform.startswith("linux"):
+    pytest.skip("skipping tests on win", allow_module_level=True)
+
+
+@pytest.mark.parametrize("val_shape", [(), (2,)])
```
Comment (czkkkkkk) [inline tests/pytorch/mock_sparse2/test_elementwise_op.py:15]: Could you check whether autograd is working? I am not sure whether `sparse_coo_tensor` has autograd.

File: python/dgl/mock_sparse2/elementwise_op.py [hunk 3]
```diff
@@ -2,20 +2,38 @@
...
+        Diagonal matrix if both :attr:`A` and :attr:`B` are diagonal matrices,
+        sparse matrix otherwise
+
+    Examples
+    --------
+    >>> row = torch.tensor([1, 0, 2])
+    >>> col = torch.tensor([0, 1, 2])
+    >>> val = torch.tensor([10, 20, 30])
+    >>> A = create_from_coo(row, col, val)
+    >>> B = diag(torch.arange(1, 4))
```
[the remark this answers is not in the document]
Comment (mufeili) [reply python/dgl/mock_sparse2/elementwise_op.py:34]: updated

File: tests/pytorch/mock_sparse2/test_elementwise_op.py (hunk 2 above)
Comment (mufeili) [reply tests/pytorch/mock_sparse2/test_elementwise_op.py:15]: It's not. I've added an item in #5022 .

**Pull request merged**